abci/precommit/precommit.go

```diff
@@ -0,0 +1,96 @@
+package precommit
+
+import (
+	"cosmossdk.io/log"
+	"github.com/cosmos/cosmos-sdk/types/module"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/skip-mev/block-sdk/block"
+)
+
+// MempoolEvictionPreCommiter is a PreCommit function that evicts txs that invalid from the app-side mempool.
+type MempoolEvictionPreCommiter struct {
+	// logger
+	logger log.Logger
+
+	// app side mempool interface
+	mempl block.Mempool
+
+	// module manager
+	mm *module.Manager
+
+	// ante handler
+	anteHandler sdk.AnteHandler
+}
+
+// NewMempoolEvictionPreCommiter returns a new MempoolEvictionPreCommiter handler.
+func NewMempoolEvictionPreCommiter(
+	logger log.Logger,
+	mempl block.Mempool,
+	mm *module.Manager,
+	anteHandler sdk.AnteHandler,
+) MempoolEvictionPreCommiter {
+	return MempoolEvictionPreCommiter{
+		logger:      logger,
+		mempl:       mempl,
+		mm:          mm,
+		anteHandler: anteHandler,
+	}
+}
+
+// PreCommit returns a PreCommit handler that wraps the default PreCommit handler and evicts invalid txs from the
+// app-side mempool
+func (m *MempoolEvictionPreCommiter) PreCommit() sdk.Precommiter {
+	return func(ctx sdk.Context) {
+		defer func() {
+			if rec := recover(); rec != nil {
+				m.logger.Error(
+					"panic in precommiter",
+					"err", rec)
+			}
+		}()
+
+		// call precommit per module
+		err := m.mm.Precommit(ctx)
+		if err != nil {
+			panic(err)
+		}
+
+		lanes, err := m.mempl.Registry(ctx)
+		if err != nil {
+			panic(err)
+		}
+
+		// add eviction hook
+		for _, lane := range lanes {
+			for iterator := lane.Select(ctx, nil); iterator != nil; iterator = iterator.Next() {
+				tx := iterator.Tx()
+
+				// run ante handler on all tx
+				_, err := m.anteHandler(ctx, tx, false)
```

Comment:
Are we sure this is the same verification as each lane? If not, we should sync them. I.e wrap lane verification in a public fn. and use that, as opposed to depending directly on anteHandlers

Reply:
hmm, to follow up on that point the base lane implements a VerifyTx interface that essentially verifies the tx according to the preferences of the lane. I would probably add that to the lane interface and use that here.

Reply:
its also not fully clear to me whether this is better than just having re-check tx called by comet. given that we didnt actually have introspection into whether the app side mempool grew/shrank in the same direction as the comet mempool, maybe it makes more sense to first loadtest with the metrics on the app side mempool. this change is probably necessary, but more introspection is probably required.

Reply:
its also unclear how performant removing txs is from the app side mempool. I imagine if the mempool is full this could potentially take on the order of a second to purge.